servant/src/Servant/API/UVerb.hs

```diff
@@ -0,0 +1,123 @@
+-- 201 created with a different body type, depending on the circumstances.  (All of this can
+-- be done with vanilla servant-server by throwing exceptions, but it can't be represented in
+-- the API types without something like `UVerb`.)
+--
+-- See <https://docs.servant.dev/en/stable/cookbook/uverb/UVerb.html> for a working example.
+module Servant.API.UVerb
+  ( UVerb,
+    Union,
+    HasStatus (StatusOf),
+    statusOf,
+    HasStatuses (Statuses, statuses),
+    WithStatus (..),
+    module Servant.API.UVerb.OpenUnion,
+    collapseUResp,
+    extractUResp,
+  )
+where
+
+import Control.Monad.Identity
+import Data.Aeson (FromJSON, ToJSON)
+import Data.Functor.Identity (Identity)
+import Data.SOP.BasicFunctors (K (K))
+import Data.SOP.Constraint (Constraint, All)
+import Data.SOP.NS (NS, cmap_NS, collapse_NS)
+import Data.Typeable (Typeable, Proxy (Proxy), cast)
+import qualified GHC.Generics as GHC
+import GHC.TypeLits (Nat)
+import Network.HTTP.Types (Status, StdMethod)
+import Servant.API.ContentTypes (MimeRender (mimeRender), MimeUnrender (mimeUnrender), NoContent)
+import Servant.API.Status (KnownStatus, statusVal)
+import Servant.API.UVerb.OpenUnion
+
+class KnownStatus (StatusOf a) => HasStatus (a :: *) where
+  type StatusOf (a :: *) :: Nat
+
+statusOf :: forall a proxy. HasStatus a => proxy a -> Status
+statusOf = const (statusVal (Proxy :: Proxy (StatusOf a)))
+
+instance KnownStatus n => HasStatus (WithStatus n a) where
+  type StatusOf (WithStatus n a) = n
+
+-- | If an API can respond with 'NoContent' we assume that this will happen
+-- with the status code 204 No Content. If this needs to be overridden,
+-- 'WithStatus' can be used.
+instance HasStatus NoContent where
+  type StatusOf NoContent = 204
+
+class HasStatuses (as :: [*]) where
+  type Statuses (as :: [*]) :: [Nat]
+  statuses :: Proxy as -> [Status]
+
+instance HasStatuses '[] where
+  type Statuses '[] = '[]
+  statuses _ = []
+
+instance (HasStatus a, HasStatuses as) => HasStatuses (a ': as) where
+  type Statuses (a ': as) = StatusOf a ': Statuses as
+  statuses _ = statusOf (Proxy :: Proxy a) : statuses (Proxy :: Proxy as)
+
+newtype WithStatus (k :: Nat) a = WithStatus a
+  deriving (Eq, Show, GHC.Generic)
+
+instance (GHC.Generic (WithStatus n a), ToJSON a) => ToJSON (WithStatus n a)
+
+instance (GHC.Generic (WithStatus n a), FromJSON a) => FromJSON (WithStatus n a)
+
+instance MimeRender ctype a => MimeRender ctype (WithStatus _status a) where
+  mimeRender contentTypeProxy (WithStatus a) = mimeRender contentTypeProxy a
+
+instance MimeUnrender ctype a => MimeUnrender ctype (WithStatus _status a) where
+  mimeUnrender contentTypeProxy input =
+    WithStatus <$> mimeUnrender contentTypeProxy input
+
+-- FUTUREWORK:
+-- @type Verb method statusCode contentTypes a = UVerb method contentTypes [WithStatus statusCode a]@
+-- no, wait, this is not the same.  this would mean people would have to use 'respond' instead
```

Comment:
Do we want to adjust this comment to be more enlightening?  

If you `newtype Verb` instead of `type Verb` you can write a `Monad` instance right? so no need for `respond` instead of `pure`

Reply:
How would the `newtype` look like?  The point of the type alias is to make `Verb` a shortcut for `UVerb`, and to be able to make `Verb`s `HasServer` instance work for `Verb` out of the box, but with `newtype`, you'd have two different types again, so how would you gain anything?

I've extended the comment a little: a024e0bab16da0c5a9ad560ae829a8c3a215a05e